File: js/json.ts
/**
 * JSON evaluation scorers for comparing and validating JSON data.
 *
 * This module provides scorers for working with JSON data:
 *
 * - **JSONDiff**: Compare JSON objects for structural and content similarity
 * - **ValidJSON**: Validate if a value is valid JSON and matches an optional schema
 *
 * ## Creating Custom JSON Scorers
 *
 * You can create custom JSON scorers by composing existing scorers or building new ones:
 *
 * @example
 * ```typescript
 * import { Scorer } from "autoevals";
 * import { JSONDiff, ValidJSON } from "autoevals/json";
 * import { EmbeddingSimilarity } from "autoevals/string";
 *
 * // Custom scorer that validates JSON schema then compares semantically
 * const myJSONScorer: Scorer<any, { schema?: any }> = async ({ output, expected, schema }) => {
 *   // First, validate both outputs against schema
 *   const outputValid = await ValidJSON({ output, schema });
 *   const expectedValid = await ValidJSON({ output: expected, schema });
 *
 *   if (outputValid.score === 0 || expectedValid.score === 0) {
 *     return {
 *       name: "CustomJSONScorer",
 *       score: 0,
 *       error: "Invalid JSON format"
 *     };
 *   }
 *
 *   // Then compare using semantic similarity for strings
 *   return JSONDiff({
 *     output,
 *     expected,
 *     stringScorer: EmbeddingSimilarity
 *   });
 * };
 *
 * // Custom scorer for specific JSON structure validation
 * const apiResponseScorer: Scorer<any, object> = async ({ output }) => {
 *   const parsed = typeof output === "string" ? JSON.parse(output) : output;
 *
 *   let score = 0;
 *   const errors: string[] = [];
 *
 *   // Check required fields
 *   if (parsed.status) score += 0.3;
 *   else errors.push("Missing status field");
 *
 *   if (parsed.data) score += 0.3;
 *   else errors.push("Missing data field");
 *
 *   // Check data structure
 *   if (parsed.data?.items && Array.isArray(parsed.data.items)) {
 *     score += 0.4;
 *   } else {
 *     errors.push("data.items must be an array");
 *   }
 *
 *   return {
 *     name: "APIResponseScorer",
 *     score: Math.min(score, 1),
 *     metadata: { errors }
 *   };
 * };
 * ```
 */

import { Scorer } from "./score";
import { NumericDiff } from "./number";
import { LevenshteinScorer } from "./string";
import Ajv, { JSONSchemaType, Schema } from "ajv";
import { makePartial, ScorerWithPartial } from "./partial";

/**
 * Compare JSON objects for structural and content similarity.
 *
 * This scorer recursively compares JSON objects, handling:
 * - Nested dictionaries and arrays
 * - String similarity using Levenshtein distance (or custom scorer)
 * - Numeric value comparison (or custom scorer)
 * - Automatic parsing of JSON strings
 *
 * @example
 * ```typescript
 * import { JSONDiff } from "autoevals";
 * import { EmbeddingSimilarity } from "autoevals/string";
 *
 * // Basic comparison
 * const result = await JSONDiff({
 *   output: {
 *     name: "John Smith",
 *     age: 30,
 *     skills: ["python", "javascript"]
 *   },
 *   expected: {
 *     name: "John A. Smith",
 *     age: 31,
 *     skills: ["python", "typescript"]
 *   }
 * });
 * console.log(result.score); // Similarity score between 0-1
 *
 * // With custom string scorer using embeddings
 * const semanticResult = await JSONDiff({
 *   output: { description: "A fast car" },
 *   expected: { description: "A quick automobile" },
 *   stringScorer: EmbeddingSimilarity
 * });
 * ```
 *
 * @param output - The JSON object or string to evaluate
 * @param expected - The expected JSON object or string to compare against
 * @param stringScorer - Optional custom scorer for string comparisons (default: LevenshteinScorer)
 * @param numberScorer - Optional custom scorer for number comparisons (default: NumericDiff)
 * @param preserveStrings - Don't attempt to parse strings as JSON (default: false)
 * @returns Score object with similarity score between 0-1
 */
export const JSONDiff: ScorerWithPartial<
  any,
  {
    stringScorer?: Scorer<string, object>;
    numberScorer?: Scorer<number, object>;
    preserveStrings?: boolean;
  }
> = makePartial(
  async ({
    output,
    expected,
    stringScorer = LevenshteinScorer,
    numberScorer = NumericDiff,
    preserveStrings = false,
  }) => {
    return {
      name: "JSONDiff",
      score: await jsonDiff(
        output,
        expected,
        stringScorer,
        numberScorer,
        preserveStrings,
      ),
    };
  },
  "JSONDiff",
);

/**
 * Validate if a value is valid JSON and optionally matches a JSON Schema.
 *
 * This scorer checks if:
 * - The input can be parsed as valid JSON (if it's a string)
 * - The parsed JSON matches an optional JSON Schema
 * - Handles both string inputs and pre-parsed JSON objects
 *
 * @example
 * ```typescript
 * import { ValidJSON } from "autoevals";
 *
 * // Basic JSON validation
 * const result1 = await ValidJSON({
 *   output: '{"name": "John", "age": 30}'
 * });
 * console.log(result1.score); // 1 (valid JSON)
 *
 * const result2 = await ValidJSON({
 *   output: '{invalid json}'
 * });
 * console.log(result2.score); // 0 (invalid JSON)
 *
 * // With schema validation
 * const schema = {
 *   type: "object",
 *   properties: {
 *     name: { type: "string" },
 *     age: { type: "number" }
 *   },
 *   required: ["name", "age"]
 * };
 *
 * const result3 = await ValidJSON({
 *   output: { name: "John", age: 30 },
 *   schema
 * });
 * console.log(result3.score); // 1 (matches schema)
 *
 * const result4 = await ValidJSON({
 *   output: { name: "John" }, // missing required "age"
 *   schema
 * });
 * console.log(result4.score); // 0 (doesn't match schema)
 * ```
 *
 * @param output - The value to validate (string or object)
 * @param schema - Optional JSON Schema to validate against (see https://json-schema.org)
 * @returns Score object with score of 1 if valid, 0 otherwise
 */
export const ValidJSON: ScorerWithPartial<any, { schema?: any }> = makePartial(
  async ({ output, schema }) => {
    return {
      name: "ValidJSON",
      score: validJSON(output, schema),
      metadata: { schema },
    };
  },
  "ValidJSON",
);

async function jsonDiff(
  o1: any,
  o2: any,
  stringScorer: Scorer<string, object>,
  numberScorer: Scorer<number, object>,
  preserveStrings: boolean,
): Promise<number | null> {
  if (!preserveStrings) {
    if (typeof o1 === "string" && validJSON(o1) === 1) {
      o1 = JSON.parse(o1);
    }
    if (typeof o2 === "string" && validJSON(o2) === 1) {
      o2 = JSON.parse(o2);
    }
  }

  if (isObject(o1) && isObject(o2)) {
    if (Object.keys(o1).length == 0 && Object.keys(o2).length == 0) {
      return 1;
    }

    const allKeys = Object.keys(
      Object.fromEntries(
        Object.keys(o1)
          .concat(Object.keys(o2))
          .map((k) => [k, true]),
      ),
    );

    // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
    const baseScores = (
      await Promise.all(
        allKeys.map((k) =>
          jsonDiff(o1[k], o2[k], stringScorer, numberScorer, preserveStrings),
        ),
      )
    ).filter((s) => s !== null) as number[];
    return baseScores.reduce((acc, s) => acc + s, 0) / baseScores.length;
  } else if (isArray(o1) && isArray(o2)) {
    if (o1.length === 0 && o2.length === 0) {
      return 1;
    }

    // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
    const baseScores = (
      await Promise.all(
        Array.from({
          length: Math.min(o1.length, o2.length),
        }).map((_, i) =>
          jsonDiff(o1[i], o2[i], stringScorer, numberScorer, preserveStrings),
        ),
      )
    ).filter((s) => s !== null) as number[];
    return (
      baseScores.reduce((acc, s) => acc + s, 0) / Math.max(o1.length, o2.length)
    );
  } else if (typeof o1 === "string" && typeof o2 === "string") {
    return (await stringScorer({ output: o1, expected: o2 })).score;
  } else if (typeof o1 === "number" && typeof o2 === "number") {
    return (await numberScorer({ output: o1, expected: o2 })).score;
  } else if (
    (o1 === null || o1 === undefined) &&
    (o2 === null || o2 === undefined)
  ) {
    return 1;
  } else if (
    o1 === null ||
    o1 === undefined ||
    o2 === null ||
    o2 === undefined
  ) {
    return 0;
  } else {
    return (
      await stringScorer({
        output: JSON.stringify(o1, replacer),
        expected: JSON.stringify(o2, replacer),
      })
    ).score;
  }
}

function isObject(value: any): value is { [key: string]: any } {
  return value instanceof Object && !(value instanceof Array);
}

function isArray(value: any): value is Array<unknown> {
  return value instanceof Array;
}

// https://gist.github.com/davidfurlong/463a83a33b70a3b6618e97ec9679e490
const replacer = (key: string, value: any) =>
  isObject(value)
    ? Object.keys(value)
        .sort()
        .reduce((sorted: { [key: string]: any }, key) => {
          sorted[key] = value[key];
          return sorted;
        }, {})
    : value;

function validJSON<T>(output: any, schema?: Schema | JSONSchemaType<T>) {
  try {
    const parsed = typeof output === "string" ? JSON.parse(output) : output;

    if (schema) {
      return validateSchema(parsed, schema);
    }
    if (isObject(parsed) || isArray(parsed)) {
      return 1;
    }
  } catch {
    // Ignore errors
  }

  return 0;
}

function validateSchema(data: any, schema: any) {
  const ajv = new Ajv();
  const validate = ajv.compile(schema);
  const valid = validate(data);
  return valid ? 1 : 0;
}

File: py/autoevals/json.py
"""JSON evaluation scorers for comparing and validating JSON data.

This module provides scorers for working with JSON data:

- JSONDiff: Compare JSON objects for structural and content similarity
  - Handles nested structures, strings, numbers
  - Customizable with different scorers for string and number comparisons
  - Can automatically parse JSON strings

- ValidJSON: Validate if a string is valid JSON and matches an optional schema
  - Validates JSON syntax
  - Optional JSON Schema validation
  - Works with both strings and parsed objects
"""

import json

from jsonschema import ValidationError, validate

from autoevals.partial import ScorerWithPartial

from .number import NumericDiff
from .score import Score, Scorer
from .string import Levenshtein


class JSONDiff(ScorerWithPartial):
    """Compare JSON objects for structural and content similarity.

    This scorer recursively compares JSON objects, handling:
    - Nested dictionaries and lists
    - String similarity using Levenshtein distance
    - Numeric value comparison
    - Automatic parsing of JSON strings

    Example:
        ```python
        import asyncio
        from openai import AsyncOpenAI
        from autoevals import JSONDiff
        from autoevals.string import EmbeddingSimilarity

        async def compare_json():
            # Initialize with async client for string comparison
            client = AsyncOpenAI()
            string_scorer = EmbeddingSimilarity(client=client)

            diff = JSONDiff(string_scorer=string_scorer)

            result = await diff.eval_async(
                output={
                    "name": "John Smith",
                    "age": 30,
                    "skills": ["python", "javascript"]
                },
                expected={
                    "name": "John A. Smith",
                    "age": 31,
                    "skills": ["python", "typescript"]
                }
            )

            print(result.score)  # Similarity score between 0-1
            print(result.metadata)  # Detailed comparison breakdown

        # Run the async evaluation
        asyncio.run(compare_json())
        ```

    Args:
        string_scorer: Optional custom scorer for string comparisons (default: Levenshtein)
        number_scorer: Optional custom scorer for number comparisons (default: NumericDiff)
        preserve_strings: Don't attempt to parse strings as JSON (default: False)

    Returns:
        Score object with:
        - score: Similarity score between 0-1
        - metadata: Detailed comparison breakdown
    """

    def __init__(self, string_scorer: Scorer = None, number_scorer: Scorer = None, preserve_strings: bool = False):
        self.string_scorer = string_scorer or Levenshtein()
        self.number_scorer = number_scorer or NumericDiff()
        self.preserve_strings = preserve_strings
        self._valid_json = ValidJSON()

    def _run_eval_sync(self, output, expected=None, **kwargs):
        return Score(name=self._name(), score=self.json_diff(output, expected))

    def json_diff(self, o1, o2):
        if not self.preserve_strings:
            if isinstance(o1, str) and self._valid_json.valid_json(o1) == 1:
                o1 = json.loads(o1)
            if isinstance(o2, str) and self._valid_json.valid_json(o2) == 1:
                o2 = json.loads(o2)

        if isinstance(o1, dict) and isinstance(o2, dict):
            if len(o1) == 0 and len(o2) == 0:
                return 1

            all_keys = set(o1.keys()).union(set(o2.keys()))
            base_scores = [self.json_diff(o1.get(k), o2.get(k)) for k in all_keys]
            base_scores = [s for s in base_scores if s is not None]
            return sum(base_scores) / len(base_scores)
        elif isinstance(o1, list) and isinstance(o2, list):
            if len(o1) == 0 and len(o2) == 0:
                return 1
            base_scores = [self.json_diff(e1, e2) for (e1, e2) in zip(o1, o2)]
            base_scores = [s for s in base_scores if s is not None]
            return sum(base_scores) / max(len(o1), len(o2))
        elif isinstance(o1, str) and isinstance(o2, str):
            return self.string_scorer.eval(o1, o2).score
        elif (isinstance(o1, int) or isinstance(o1, float)) and (isinstance(o2, int) or isinstance(o2, float)):
            return self.number_scorer.eval(o1, o2).score
        elif o1 is None and o2 is None:
            return 1
        elif o1 is None or o2 is None:
            return 0
        else:
            kwargs = {"separators": (",", ":"), "sort_keys": True}
            return self.string_scorer.eval(json.dumps(o1, **kwargs), json.dumps(o2, **kwargs)).score


class ValidJSON(ScorerWithPartial):
    """Validate if a string is valid JSON and optionally matches a schema.

    This scorer checks if:
    - The input can be parsed as valid JSON
    - The parsed JSON matches an optional JSON Schema
    - Handles both string inputs and pre-parsed JSON objects

    Example:
        ```python
        import asyncio
        from autoevals import ValidJSON

        async def validate_json():
            # Define a schema to validate against
            schema = {
                "type": "object",
                "properties": {
                    "name": {"type": "string"},
                    "age": {"type": "number"},
                    "skills": {
                        "type": "array",
                        "items": {"type": "string"}
                    }
                },
                "required": ["name", "age"]
            }

            validator = ValidJSON(schema=schema)

            result = await validator.eval_async(
                output='''
                {
                    "name": "John Smith",
                    "age": 30,
                    "skills": ["python", "javascript"]
                }
                '''
            )

            print(result.score)  # 1 if valid, 0 if invalid
            print(result.metadata)  # Validation details or error messages

        # Run the async validation
        asyncio.run(validate_json())
        ```

    Args:
        schema: Optional JSON Schema to validate against

    Returns:
        Score object with:
        - score: 1 if valid JSON (and matches schema if provided), 0 otherwise
        - metadata: Validation details or error messages
    """

    def __init__(self, schema=None):
        self.schema = schema

    def _run_eval_sync(self, output, schema=None, **kwargs):
        return Score(name=self._name(), score=self.valid_json(output, schema))

    def valid_json(self, output, schema=None):
        try:
            parsed = json.loads(output) if isinstance(output, str) else output

            if schema is not None:
                validate(parsed, schema)
                return 1

            if isinstance(parsed, dict) or isinstance(parsed, list):
                return 1

        except (json.JSONDecodeError, ValidationError):
            pass

        return 0


__all__ = ["JSONDiff", "ValidJSON"]
